# Part 1. DATA CLEANING AND EXPLORATION

### Step 1. Upload the Dataset

In [1]:
from google.colab import files
file = files.upload()

Saving Superstore.csv to Superstore.csv


### Step 2. Import Required Libraries

In [2]:
#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

#Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)


print("✅ Libraries imported successfully!")

✅ Libraries imported successfully!


### Step 3: Load the Data

In [3]:
df = pd.read_csv("Superstore.csv", encoding = "latin-1")
print("Dataset Loaded Sucessfully")
print(f"Shape: {df.shape[0]} rows, {df.shape[1]} columns")

Dataset Loaded Sucessfully
Shape: 9994 rows, 21 columns


The super store sales dataset has total 9994 rows and 21 variables.

### Step 4: Initital Data Exploration

In [4]:
print("="*80)
print("FIRST 5 ROWS")
print("="*80)
display(df.head())

print("\n" + "="*80)
print("DATASET INFORMATION")
print("="*80)
df.info()

print("\n" + "="*80)
print("STATISTICAL SUMMARY")
print("="*80)
display(df.describe())

print("\n" + "="*80)
print("COLUMN NAMES")
print("="*80)
display(df.columns.tolist())



FIRST 5 ROWS


Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2013-152156  09-11-2013  12-11-2013    Second Class    CG-12520   
1       2  CA-2013-152156  09-11-2013  12-11-2013    Second Class    CG-12520   
2       3  CA-2013-138688  13-06-2013  17-06-2013    Second Class    DV-13045   
3       4  US-2012-108966  11-10-2012  18-10-2012  Standard Class    SO-20335   
4       5  US-2012-108966  11-10-2012  18-10-2012  Standard Class    SO-20335   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region       Product ID         Category Sub-Category  \
0        42420  South  FUR-BO-10001798        Furniture    Bookcases   
1        42420  South  FUR-CH-10000454        Furniture       Chairs   
2        90036   West  OFF-LA-10000240  Office Supplies       Labels   
3        33311  South  FUR-TA-10000577        Furniture       Tables   
4        33311  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164


DATASET INFORMATION
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   

Row ID   Postal Code         Sales     Quantity     Discount  \
count  9994.000000   9994.000000   9994.000000  9994.000000  9994.000000   
mean   4997.500000  55190.379428    229.858001     3.789574     0.156203   
std    2885.163629  32063.693350    623.245101     2.225110     0.206452   
min       1.000000   1040.000000      0.444000     1.000000     0.000000   
25%    2499.250000  23223.000000     17.280000     2.000000     0.000000   
50%    4997.500000  56430.500000     54.490000     3.000000     0.200000   
75%    7495.750000  90008.000000    209.940000     5.000000     0.200000   
max    9994.000000  99301.000000  22638.480000    14.000000     0.800000   

            Profit  
count  9994.000000  
mean     28.656896  
std     234.260108  
min   -6599.978000  
25%       1.728750  
50%       8.666500  
75%      29.364000  
max    8399.976000


COLUMN NAMES


['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit']

The above code cell is used to print first 5 rows, summary of the data, info of the data, and column names in super store sales dataset.

### Step 5: Check Data Quality

In [5]:
print("\n" + "="*80)
print("MISSING VALUES")
print("=" * 80)
missing = df.isnull().sum()
missing_pct = df.isnull().sum() / len(df) * 100
missing_df = pd.DataFrame({'Column':missing.index,
                           'Missing Count':missing.values,
                           'Percentage':missing_pct.values})
missing_df = missing_df[missing_df['Missing Count'] > 0].sort_values(by = "Missing Count", ascending = False)
print(missing_df if len(missing_df) > 0 else "✅ No missing values found!")

# Check for duplicates
print("\n"+ "=" * 80)
print("DUPLICATE ROWS")
print("=" * 80)
duplicates = df.duplicated().sum()
print(f"NUmber of duplicate rows: {duplicates}" )


# Check unique value for key categorical columns
print("\n" + "=" * 80)
print("UNIQUE VALUES IN CATEGORICAL COLUMNS")
print("=" * 80)
categorical_cols = df.select_dtypes(include = ["object"]).columns
for col in categorical_cols:
    print(f"{col}: {df[col].nunique()} unique values")
    if df[col].nunique() < 20:
        print(f"Values: {df[col].unique()}")
    print()


MISSING VALUES
✅ No missing values found!

DUPLICATE ROWS
NUmber of duplicate rows: 0

UNIQUE VALUES IN CATEGORICAL COLUMNS
Order ID: 5009 unique values

Order Date: 1238 unique values

Ship Date: 1334 unique values

Ship Mode: 4 unique values
Values: ['Second Class' 'Standard Class' 'First Class' 'Same Day']

Customer ID: 793 unique values

Customer Name: 793 unique values

Segment: 3 unique values
Values: ['Consumer' 'Corporate' 'Home Office']

Country: 1 unique values
Values: ['United States']

City: 531 unique values

State: 49 unique values

Region: 4 unique values
Values: ['South' 'West' 'Central' 'East']

Product ID: 1862 unique values

Category: 3 unique values
Values: ['Furniture' 'Office Supplies' 'Technology']

Sub-Category: 17 unique values
Values: ['Bookcases' 'Chairs' 'Labels' 'Tables' 'Storage' 'Furnishings' 'Art'
 'Phones' 'Binders' 'Appliances' 'Paper' 'Accessories' 'Envelopes'
 'Fasteners' 'Supplies' 'Machines' 'Copiers']

Product Name: 1841 unique values



The output showed that there is no missing value or duplicate rows in the dataset.

### Step 6: Initial Business Questions

In [6]:
# Quick Business Insights
print("\n" + "=" * 80)
print("QUICK BUSINESS INSIGHTS")
print("" + "=" * 80)

#Total metrics
print(f"💰 Total Sales: ${df['Sales'].sum():,.2f}")
print(f"📈 Total Profit: ${df['Profit'].sum():,.2f}")
print(f"📦 Total Orders:, {df['Order ID'].nunique():,}")
print(f"👥 Total Customers: {df['Customer ID'].nunique():,}")
print(f"🏷️ Total Products: {df['Product ID'].nunique():,}")


#Profit Margin
overall_profit_margin = (df['Profit'].sum() / df['Sales'].sum()) * 100
print(f"💹 Overall Profit Margin: {overall_profit_margin:.2f}%")

#Date Range
if 'Order Date' in df.columns:
    print(f"📅 Date Range: {df['Order Date'].min()} to {df['Order Date'].max()}")


#Nagative Profits
negative_profits = (df['Profit'] < 0).sum()
negative_pct = (negative_profits / len(df)) * 100
print(f"⚠️ Unprofitable Orders: {negative_profits} ({negative_pct:.2f}%) ")


QUICK BUSINESS INSIGHTS
💰 Total Sales: $2,297,200.86
📈 Total Profit: $286,397.02
📦 Total Orders:, 5,009
👥 Total Customers: 793
🏷️ Total Products: 1,862
💹 Overall Profit Margin: 12.47%
📅 Date Range: 01-01-2014 to 31-12-2014
⚠️ Unprofitable Orders: 1871 (18.72%) 


### Step 7. Fix Date Columns

In [7]:
print("="*80)
print("CONVERTING DATE COLUMNS")
print("="*80)
df['Order Date'] = pd.to_datetime(df['Order Date'], format = "%d-%m-%Y")
df['Ship Date'] = pd.to_datetime(df['Ship Date'], format = "%d-%m-%Y")
print("✅ Date columns converted successfully")
print(f"Order Date Range: {df['Order Date'].min()} to {df['Order Date'].max()}")
print(f"Ship Date Range: {df['Ship Date'].min()} to {df['Ship Date'].max()}")

print("\nData types after conversion")
print(df.loc[:,['Order Date', 'Ship Date']].dtypes)


CONVERTING DATE COLUMNS
✅ Date columns converted successfully
Order Date Range: 2011-01-04 00:00:00 to 2014-12-31 00:00:00
Ship Date Range: 2011-01-08 00:00:00 to 2015-01-06 00:00:00

Data types after conversion
Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object


### Step 8: Create Time-Based Features

In [8]:
print("\n" + "=" * 80)
print("CREATING TIME-BASED FEATURES")
print("="*80)
df["Year"] = df["Order Date"].dt.year
df["Month"] = df["Order Date"].dt.month
df["Month Name"] = df["Order Date"].dt.month_name()
df["Quarter"] = df["Order Date"].dt.quarter
df["Day of Week"] = df["Order Date"].dt.day_name()
df["Week of Year"] = df["Order Date"].dt.isocalendar().week

#Shipping time in days
df["Days to Ship"] = (df["Ship Date"] - df["Order Date"]).dt.days
print("✅ Time features created")
print("\nNew columns added:")
print("- Year, Month, Month Name, Quarter")
print("- Day of Week, Week of Year")
print("- Days to Ship")

#Quick Check
print("\nYear distribution: ")
print(df["Year"].value_counts().sort_index())

print("\nAverage days to ship by Ship Mode:")
print(df.groupby('Ship Mode')['Days to Ship'].mean().sort_values())


CREATING TIME-BASED FEATURES
✅ Time features created

New columns added:
- Year, Month, Month Name, Quarter
- Day of Week, Week of Year
- Days to Ship

Year distribution: 
Year
2011    1993
2012    2102
2013    2580
2014    3319
Name: count, dtype: int64

Average days to ship by Ship Mode:
Ship Mode
Same Day          0.044199
First Class       2.182055
Second Class      3.240103
Standard Class    5.007205
Name: Days to Ship, dtype: float64


### Step 9: Create Financial Metrics

In [9]:
#Calculate derived financial metrics

print("\n" + "="*80)
print("CREATING FINANCIAL METRICS")
print("="*80)

# Profit margin percentages
df['Profit Margin %'] = df['Profit'] / df['Sales'] * 100

# Revenue Per Unit
df['Revenue per Unit'] = df['Sales'] / df['Quantity']

# Cost (Sales - Profit)
df['Cost'] = df['Sales'] - df['Profit']

# Discount Amount in dollars
df['Discount Amount'] = df['Sales'] * df['Discount']

# Revenue without discount
df['Original Price'] = df['Sales'] / (1 - df['Discount'])
df.loc[df['Discount'] == 0, 'Original Price'] = df['Sales']

# Profitability flage
df['Is Profitable'] = df['Profit'] > 0

print("✅ Financial metrics created!")
print("\nNew columns added:")
print("- Profit Marging %")
print("- Revenue per Unit")
print("- Cost")
print("- Discount Amount")
print("- Original Price")
print("- Is Profitable (Boolean)")

#Summary stats
print("\n📊 Profit Margin Distribution")
print(df["Profit Margin %"].describe())

print("\n💰 Average Metrics by Profitability")
profitable_comparison = df.groupby('Is Profitable').agg({
    'Sales': 'mean', 'Profit': 'mean', 'Discount':'mean', 'Profit Margin %':'mean'
}).round(2)
display(profitable_comparison)


CREATING FINANCIAL METRICS
✅ Financial metrics created!

New columns added:
- Profit Marging %
- Revenue per Unit
- Cost
- Discount Amount
- Original Price
- Is Profitable (Boolean)

📊 Profit Margin Distribution
count    9994.000000
mean       12.031393
std        46.675435
min      -275.000000
25%         7.500000
50%        27.000000
75%        36.250000
max        50.000000
Name: Profit Margin %, dtype: float64

💰 Average Metrics by Profitability


Sales  Profit  Discount  Profit Margin %
Is Profitable                                           
False          256.40  -80.65      0.47           -60.36
True           223.48   54.92      0.08            29.42

### Step 10: Create Customer & Product Aggregations

In [10]:
#Customer-level metrics
print("\n" + "=" * 80)
print("CREATING CUSTOMER METRICS")
print("="*80)

# Orders per customer
customer_orders = df.groupby("Customer ID").size().reset_index(name = 'Total Orders')
df = df.merge(customer_orders, on = 'Customer ID', how = 'left')

#Customer lifetime value
customer_value = df.groupby("Customer ID")["Sales"].sum().reset_index(name = "Customer Lifetime Value")
df = df.merge(customer_value, on = "Customer ID", how = "left")

#Customer category (based on number of orders)
df["Customer Type"] = df["Total Orders"].apply(
    lambda x: "One-time" if x == 1
    else "Occasional" if x <= 3
    else "Regular" if x <= 7
    else " Loyal"
)

print("✅ Customer metrics created")
print("\nCustomer Type Distribution:")
print(df.groupby("Customer Type")["Customer ID"].nunique())

print("\n💎 Top 5 Customers by Lifetime Value")
top_customers = df.groupby(["Customer ID", 'Customer Name', "Customer Lifetime Value"]).size().reset_index()
top_customers = top_customers.drop_duplicates('Customer ID').sort_values(by = "Customer Lifetime Value", ascending = False).head()
display(top_customers[['Customer Name', 'Customer Lifetime Value']])


CREATING CUSTOMER METRICS
✅ Customer metrics created

Customer Type Distribution:
Customer Type
 Loyal        622
Occasional     28
One-time        5
Regular       138
Name: Customer ID, dtype: int64

💎 Top 5 Customers by Lifetime Value


Customer Name  Customer Lifetime Value
700    Sean Miller                25043.050
741   Tamara Chand                19052.218
621   Raymond Buch                15117.339
730   Tom Ashbrook                14595.620
6    Adrian Barton                14473.571

### Step 11: Creating Category Performance Metrics

In [11]:
# Product/Category performance
print("\n" + "="*80)
print("CREATING PRODUCT PERFORMANCE METRICS")
print("="*80)

#Category profit margins
category_margins = df.groupby("Category").agg({
    "Sales":"sum",
    "Profit":"sum",
    "Quantity":"sum"
}).round(2)
category_margins['Profit Margin %'] = (category_margins['Profit'] / category_margins['Sales'] * 100).round(2)
print("📊 Category Performance:")
print(category_margins)

#Sub-Category Profit Margin
subcategory_margins = df.groupby("Sub-Category").agg({
    'Sales':'sum',
    'Profit':'sum'
}).round(2)
subcategory_margins['Profit Margin %'] = (subcategory_margins['Profit'] / subcategory_margins['Sales'] * 100).round(2)
subcategory_margins = subcategory_margins.sort_values(by = 'Profit Margin %', ascending = False)

print("\n🏆 Top 5 Most Profitable Sub-Categories")
print(subcategory_margins.head())

print("\n⚠️ Bottom 5 Sub-Categories (Least Profitable):")
print(subcategory_margins.tail())


CREATING PRODUCT PERFORMANCE METRICS
📊 Category Performance:
                     Sales     Profit  Quantity  Profit Margin %
Category                                                        
Furniture        741999.80   18451.27      8028             2.49
Office Supplies  719047.03  122490.80     22906            17.04
Technology       836154.03  145454.95      6939            17.40

🏆 Top 5 Most Profitable Sub-Categories
                  Sales    Profit  Profit Margin %
Sub-Category                                      
Labels         12486.31   5546.25            44.42
Paper          78479.21  34053.57            43.39
Envelopes      16476.40   6964.18            42.27
Copiers       149528.03  55617.82            37.20
Fasteners       3024.28    949.52            31.40

⚠️ Bottom 5 Sub-Categories (Least Profitable):
                  Sales    Profit  Profit Margin %
Sub-Category                                      
Chairs        328449.10  26590.17             8.10
Machines      1

### Step 12: Analyze Discount Impact

In [12]:
# Discount analysis
print("\n" + "="*80)
print("DISCOUNT IMPACT ANALYSIS")
print("="*80)

# Create discount bins
df['Discount Range'] = pd.cut(df['Discount'], bins = [-0.01, 0, 0.1, 0.2, 0.3, 1.0],
                              labels = ["No Discount", "1-10%", "11-20%", "21-30%", "30%+"])

# Analyze profitability by discount range
discount_analysis = df.groupby('Discount Range').agg({
    'Order ID': 'count',
    'Sales': 'mean',
    'Profit': 'mean',
    'Profit Margin %': 'mean',
    'Is Profitable': lambda x: (x.sum() / len(x) * 100)
}).round(2)
discount_analysis.columns = ['Order Count', 'Avg Sales', 'Avg Profit', 'Avg Margin%', '% Profitable']
print("💸 Discount Impact on Profitability:")
print(discount_analysis)


# Key Insight
high_discount_unprofitable = df[(df['Discount'] > 0.3) & (df['Profit'] < 0)].shape[0]
print(f"\n⚠️ Critical Finding: {high_discount_unprofitable} orders with 30%+ discount are unprofitable")


DISCOUNT IMPACT ANALYSIS
💸 Discount Impact on Profitability:
                Order Count  Avg Sales  Avg Profit  Avg Margin%  % Profitable
Discount Range                                                               
No Discount            4798     226.74       66.90        34.02         99.37
1-10%                    94     578.40       96.06        15.58         95.74
11-20%                 3709     213.58       24.74        17.48         85.41
21-30%                  227     454.74      -45.68       -11.55          2.64
30%+                   1166     222.59     -107.21       -91.47          2.23

⚠️ Critical Finding: 1140 orders with 30%+ discount are unprofitable


### Step 13: Regional & Segment Analysis

In [13]:
#Regional Performance
print("\n" + "="*80)
print("REGIONAL & SEGMENT ANALYSIS")
print("="*80)

#Region Performance
region_performance = df.groupby("Region").agg({
    "Sales":"sum",
    "Profit":"sum",
    "Order ID": "nunique"
}).round(2)
region_performance["Prfoit Margin %"] = region_performance["Profit"] / region_performance["Sales"] * 100
region_performance.columns = ["Total Sales", "Total Profit", "Order Count", "Profit Margin %"]
print("🗺️ Regional Performance:")
print(region_performance.sort_values("Total Profit", ascending = False))

# Segment performance
segment_performance = df.groupby("Segment").agg({
    "Sales": "sum",
    "Profit": 'sum',
    "Customer ID": "nunique"
}).round(2)
segment_performance["Profit Margin %"] = (segment_performance["Profit"] / segment_performance["Sales"] * 100).round(2)
segment_performance.columns = ["Total Sales", "Total Profit", "Customer Count", "Profit Margin %"]

print("\n👥 Segment Performance:")
print(segment_performance.sort_values('Total Profit', ascending = False))


REGIONAL & SEGMENT ANALYSIS
🗺️ Regional Performance:
         Total Sales  Total Profit  Order Count  Profit Margin %
Region                                                          
West       725457.82     108418.45         1611        14.944832
East       678781.24      91522.78         1401        13.483399
South      391721.90      46749.43          822        11.934342
Central    501239.89      39706.36         1175         7.921628

👥 Segment Performance:
             Total Sales  Total Profit  Customer Count  Profit Margin %
Segment                                                                
Consumer      1161401.34     134119.21             409            11.55
Corporate      706146.37      91979.13             236            13.03
Home Office    429653.15      60298.68             148            14.03


### Step 14: Summary of Data Preparation

In [14]:
# Final data check
print("\n" + "=" * 80)
print("DATA PREPARATION COMPLETE ✅")
print("=" * 80)

print(f"Original columns: 21")
print(f"New columns added: {len(df.columns) - 21}")
print(f"Total columns now: {len(df.columns)}")

print("\n📋 All columns:")
for i, col in enumerate(df.columns, 1):
    print(f"{i}. {col}")

print(f"\n✅ Dataset shape: {df.shape}")
print("✅ Ready for visualization and dashboard creation!")


DATA PREPARATION COMPLETE ✅
Original columns: 21
New columns added: 17
Total columns now: 38

📋 All columns:
1. Row ID
2. Order ID
3. Order Date
4. Ship Date
5. Ship Mode
6. Customer ID
7. Customer Name
8. Segment
9. Country
10. City
11. State
12. Postal Code
13. Region
14. Product ID
15. Category
16. Sub-Category
17. Product Name
18. Sales
19. Quantity
20. Discount
21. Profit
22. Year
23. Month
24. Month Name
25. Quarter
26. Day of Week
27. Week of Year
28. Days to Ship
29. Profit Margin %
30. Revenue per Unit
31. Cost
32. Discount Amount
33. Original Price
34. Is Profitable
35. Total Orders
36. Customer Lifetime Value
37. Customer Type
38. Discount Range

✅ Dataset shape: (9994, 38)
✅ Ready for visualization and dashboard creation!


### Step 15: Save Cleaned Data

In [15]:
# Save cleaned dataset
print("\n" + "=" * 80)
print("SAVING CLEANED DATA")
print("=" * 80)

# Save to CSV
df.to_csv('superstore_cleaned.csv', index=False)
print("✅ Cleaned data saved as 'superstore_cleaned.csv'")

# Also save to Google Drive (optional)
try:
    df.to_csv('/content/drive/MyDrive/superstore_cleaned.csv', index=False)
    print("✅ Also saved to Google Drive")
except:
    print("ℹ️ Google Drive not mounted - only local save")

print("\n🎉 Day 1 Complete!")
print("Next: Create visualizations and build dashboard")


SAVING CLEANED DATA
✅ Cleaned data saved as 'superstore_cleaned.csv'
ℹ️ Google Drive not mounted - only local save

🎉 Day 1 Complete!
Next: Create visualizations and build dashboard


---

## 📝 Day 1 Summary: Key Findings

### ✅ Data Quality
- **9,994 transactions** from 2014-2017
- **100% complete** - no missing values
- **0 duplicates** - clean dataset

### 💡 Critical Business Insights

**1. Profitability Crisis:**
- 18.72% of orders are UNPROFITABLE (1,871 orders)
- Overall profit margin: only 12.47%
- High discounts (30%+) correlate with losses

**2. Discount Problem:**
- Orders with 30%+ discount: mostly unprofitable
- Need discount policy revision

**3. Category Performance:**
- Technology: Highest margin
- Furniture: Lowest margin (likely the problem)
- Tables sub-category: Major loss driver

**4. Customer Insights:**
- 793 customers, avg 6.3 orders each
- Loyal customers (8+ orders) drive most profit
- Customer lifetime value varies significantly

**5. Regional Differences:**
- [Will analyze in visualizations]

### 🎯 Next Steps (Day 2)
1. Create 5 key visualizations
2. Build interactive Streamlit dashboard
3. Generate actionable recommendations

---

# Day2: DEEP INSIGHTS & DATA VISUALIZATION

## PART 1: UNPRECEDENTED INSIGHTS

### Finding 1: The Profit Cliff - Discovering the Discount Breaking Point

In [16]:
# Create granular discount bins
df['Discount Granular'] = pd.cut(df['Discount'],
                                 bins = np.arange(0, 0.85, 0.05),
                                 labels = [f'{i} - {i + 5}%' for i in range(0, 80, 5)])
# Calaculate profitability rate by exact discount range
profit_cliff = df.groupby('Discount Granular').agg({
    'Is Profitable': ['sum', 'count', 'mean'],
    'Profit': 'sum',
    'Sales': 'sum'
}).reset_index()
profit_cliff.columns = ['Discount Range', 'Profitable Orders', 'Total Orders',
                        'Profit Rate', 'Total Profit', 'Total Sales']
profit_cliff['Profit Rate %'] = (profit_cliff['Profit Rate'] * 100).round(2)
profit_cliff['Avg Profit per Order'] = (profit_cliff['Total Profit'] / profit_cliff['Total Orders']).round(2)
print("="*80)
print("🎯 THE PROFIT CLIFF ANALYSIS")
print("="*80)
print(profit_cliff[['Discount Range', 'Total Orders', 'Profit Rate %', 'Avg Profit per Order']].fillna(0, axis = 1))

tipping_point = profit_cliff[profit_cliff['Profit Rate %'] < 50].iloc[0] if len(profit_cliff[profit_cliff['Profit Rate %'] < 50]) > 0 else None
if tipping_point is not None:
    print(f"\n🚨 CRITICAL FINDING")
    print(f"   At {tipping_point['Discount Range']} discount range: ")
    print(f"   - Profitability drops below 50%")
    print(f"   - Average profit per order: ${tipping_point['Avg Profit per Order']}")
    print(f"\n💡 RECOMMENDATION: Cap discounts at {tipping_point['Discount Range'].split('-')[0]}%")



🎯 THE PROFIT CLIFF ANALYSIS
   Discount Range Total Orders Profit Rate % Avg Profit per Order
0          0 - 5%            0             0                    0
1         5 - 10%           94         95.74                96.06
2        10 - 15%           52         67.31                27.29
3        15 - 20%         3657         85.67                 24.7
4        20 - 25%            0             0                    0
5        25 - 30%          227          2.64               -45.68
6        30 - 35%           27           0.0               -88.56
7        35 - 40%          206         12.62              -111.93
8        40 - 45%           11           0.0              -226.65
9        45 - 50%           66           0.0               -310.7
10       50 - 55%            0             0                    0
11       55 - 60%          138           0.0               -43.08
12       60 - 65%            0             0                    0
13       65 - 70%          418           0.0    

In [17]:
# Visualize the cliff
fig = go.Figure()
fig.add_trace(go.Scatter(
    x = profit_cliff['Discount Range'],
    y = profit_cliff['Profit Rate %'],
    mode = 'lines+markers',
    name = 'Profitability Rate',
    line = dict(color = 'red', width = 3),
    marker = dict(size = 10)
))
fig.add_hline(y = 50, line_dash = "dash", line_color = "orange", annotation_text = "50% Threshold")
fig.update_layout(title = "🎯 The Profit Cliff: Where Dicounts Kill Profitability",
                  xaxis_title = "Discount Range",
                  yaxis_title = "Profitable Orders %",
                  template = "plotly_white", height = 500)
fig.show()

### Finding #2. Financial Impact Analysis

In [18]:
print("\n" + "="*80)
print("💸 FINANCIAL IMPACT ANALYSIS - THE REAL COST OF BAD DISCOUNTS")

# Calculate total losses by discount range
high_discount_losses = df[df['Discount'] >= 0.25].groupby(
    pd.cut(df[df['Discount'] >= 0.25]['Discount'],
        bins = np.arange(0.25, 0.85, 0.05),
           labels = [f'{i}-{i+5}%' for i in range(25, 80, 5)])
).agg({
    'Profit': ['sum', 'count', 'mean'],
    'Sales':'sum',
}).round(2)
high_discount_losses.columns = ['Total Loss', 'Order Count', 'Avg Loss', 'Total Sales']
high_discount_losses  = high_discount_losses.reset_index()

print("\n 🔴 HIGH DISCOUNT DAMAGE REPORT (25%+ discounts):")
print(high_discount_losses)

# Calculate TOTAL annual loss
total_loss_25plus = df[df['Discount'] >= 0.25]['Profit'].sum()
total_orders_25plus = len(df[df['Discount'] >= 0.25])
total_sales_25plus = df[df['Discount'] >= 0.25]['Sales'].sum()
print('\n' + '='*80)
print('🚨 EXECUTIVE SUMMARY: THE 25%+ DISCOUNT DISASTER')
print("="*80)
print(f"Total orders with 25%+ count: {total_orders_25plus:,}")
print(f"Total sales from these orders: {total_sales_25plus:,}")
print(f"Total Loss from these orders: {total_loss_25plus:,}")
print(f"Average Loss per order: ${total_loss_25plus / total_orders_25plus:,.2f}")

# What if we capped at 20%?
safe_zone_margin = df[df['Discount'] <= 0.20]['Profit'].sum()/df[df['Discount'] <= 0.20]['Sales'].sum()
potential_profit_if_capped = total_sales_25plus * safe_zone_margin
print("\n" + "="*80)
print("💡 IF WE HAD CAPPED DISCOUNTS AT 20%:")
print("="*80)
print(f"Expected profit margin:{safe_zone_margin * 100:.2f}%")
print(f"Projected porfit: ${potential_profit_if_capped:,.2f}")
print(f"Actual loss: ${total_loss_25plus}")
print(f"\n💰 TOTAL OPPORTUNITY LOST: ${potential_profit_if_capped - total_loss_25plus}")

# Annual extrapolation (4 years of date)
years = df['Order Date'].dt.year.nunique()
annual_opportunity = (potential_profit_if_capped - total_loss_25plus) / years
print(f"\n🎯 ANNUAL RECOVERABLE PROFIT: ${annual_opportunity:,.2f}")
print("\n"+"="*80)
print("🔍 WHO IS APPROVING THESE DISCOUNTS?")
print("="*100)

# Analyze extreme discounts by segment and category
extreme_discounts = df[df["Discount"] >= 0.40].groupby(["Segment", "Category"]).agg({
    "Order ID": "count",
    "Profit": "sum",
    "Sales": "sum",
    "Discount": "mean"
}).round(2)
extreme_discounts.columns = ["Order Count", "Total Loss", "Total Sales", "Avg Discount"]
extreme_discounts = extreme_discounts.sort_values(by = "Total Loss")
print("\n⚠️ EXTREME DISCOUNTS (40%+) BY SEGMENT & CATEGORY: ")
print(extreme_discounts)

# WHich customers are getting these crazy discounts?
crazy_customer_discounts = df[df["Discount"] >= 0.40].groupby("Customer ID").agg({
    "Order ID": "count",
    "Profit": "sum",
    "Sales": "sum",
    "Discount": "mean",
    "Customer Name": "first"
}).sort_values(by = "Profit").head(10)
crazy_customer_discounts.columns = ["Orders with 40%+ Discount", "Total Loss", "Total Sales", "Avg Discount", "Customer Name"]
print("\n TOP 10 LOSS-MAKING CUSTOMERS (Due to execessive discounts):")
print(crazy_customer_discounts)



💸 FINANCIAL IMPACT ANALYSIS - THE REAL COST OF BAD DISCOUNTS

 🔴 HIGH DISCOUNT DAMAGE REPORT (25%+ discounts):
   Discount  Total Loss  Order Count  Avg Loss  Total Sales
0    25-30%   -10369.28          227    -45.68    103226.66
1    30-35%    -2391.14           27    -88.56     14493.46
2    35-40%        0.00            0       NaN         0.00
3    40-45%   -23057.05          206   -111.93    116417.78
4    45-50%    -2493.11           11   -226.65      5484.97
5    50-55%   -20506.43           66   -310.70     58918.54
6    55-60%        0.00            0       NaN         0.00
7    60-65%    -5944.66          138    -43.08      6644.70
8    65-70%   -40075.36          418    -95.87     40620.28
9    70-75%        0.00            0       NaN         0.00
10   75-80%        0.00            0       NaN         0.00

🚨 EXECUTIVE SUMMARY: THE 25%+ DISCOUNT DISASTER
Total orders with 25%+ count: 1,393
Total sales from these orders: 362,770.1498
Total Loss from these orders: -135,376.

In [19]:
# Create the most impactful chart you'll ever make

# Prepare data for visualization
viz_data = profit_cliff.copy()
viz_data['Color'] = viz_data['Profit Rate %'].apply(
    lambda x: 'green' if x > 80 else 'orange' if x > 50 else 'red'
)

fig = go.Figure()

# Bar chart for profitability rate
fig.add_trace(go.Bar(
    x=viz_data['Discount Range'],
    y=viz_data['Profit Rate %'],
    name='Profitability Rate %',
    marker_color=viz_data['Color'],
    text=viz_data['Profit Rate %'].round(1),
    textposition='outside',
    hovertemplate='<b>%{x}</b><br>Profit Rate: %{y:.1f}%<br>Orders: %{customdata}<extra></extra>',
    customdata=viz_data['Total Orders']
))

# Add 50% threshold line
fig.add_hline(y=50, line_dash="dash", line_color="red",
              annotation_text="50% Survival Line",
              annotation_position="right")

# Add 80% excellence line
fig.add_hline(y=80, line_dash="dash", line_color="green",
              annotation_text="80% Excellence Zone",
              annotation_position="right")

# Highlight the death zone
fig.add_vrect(x0=4.5, x1=15.5, fillcolor="red", opacity=0.1,
              layer="below", line_width=0,
              annotation_text="DEATH ZONE", annotation_position="top left")

fig.update_layout(
    title={
        'text': "🚨 The Profit Cliff: Where Discounts Kill Your Business<br><sub>97% of orders with 25%+ discounts are UNPROFITABLE</sub>",
        'x': 0.5,
        'xanchor': 'center'
    },
    xaxis_title="Discount Range",
    yaxis_title="Profitability Rate (%)",
    template='plotly_white',
    height=600,
    showlegend=False,
    font=dict(size=12)
)

fig.show()

# Second chart: Financial impact
fig2 = go.Figure()

impact_data = df.groupby(
    pd.cut(df['Discount'],
           bins=[0, 0.20, 0.25, 1.0],
           labels=['Safe (0-20%)', 'Warning (20-25%)', 'Danger (25%+)'])
).agg({
    'Profit': 'sum',
    'Sales': 'sum',
    'Order ID': 'count'
}).reset_index()

impact_data.columns = ['Zone', 'Total Profit', 'Total Sales', 'Order Count']

fig2.add_trace(go.Bar(
    x=impact_data['Zone'],
    y=impact_data['Total Profit'],
    marker_color=['green', 'orange', 'red'],
    text=[f"${x:,.0f}" for x in impact_data['Total Profit']],
    textposition='outside'
))

fig2.update_layout(
    title="💰 Profit Distribution by Discount Zone",
    xaxis_title="Discount Zone",
    yaxis_title="Total Profit ($)",
    template='plotly_white',
    height=500
)

fig2.show()

### Finding #3: First Purchase Category Predicts Customer Value

In [27]:
# Get first purchase category for each customer
first_purchase = df.sort_values(['Customer ID', 'Order Date']).groupby('Customer ID').first().reset_index()
first_purchase_category = first_purchase[['Customer ID', 'Category']].rename(columns = {'Category':'First Purchase Category'})

# Merge back to main dataset
customer_analysis = df.groupby('Customer ID', as_index = False).agg({
    'Sales': 'sum',
    'Profit': 'sum',
    'Order ID': 'nunique',
    'Discount' : 'mean'
})
customer_analysis.columns = ['Customer ID', 'Total Sales', 'Total Profit', 'Total Orders', 'Avg Discount']
customer_analysis = customer_analysis.merge(first_purchase_category, on = 'Customer ID')

# Compare by first purchase category
first_category_impact = customer_analysis.groupby('First Purchase Category').agg({
    'Total Sales': 'mean',
    'Total Profit': 'mean',
    'Total Orders': 'mean',
    'Avg Discount': 'mean',
    'Customer ID': 'count'
}).reset_index()
first_category_impact.columns = ['First Category', 'Avg Lifetime Sales', 'Avg Lifetime Profit', 'Avg Orders', 'Avg Discount', 'Customer Count']
print("\n" + "="*80)
print("🎯 FIRST PURCHASE CATEGORY IMPACT ON CUSTOMER LIFETIME VALUE")
print(first_category_impact)

#Calculate Impact
furniture_customers = first_category_impact.loc[first_category_impact['First Category'] == 'Furniture', 'Avg Lifetime Profit'].values[0]
tech_customers = first_category_impact.loc[first_category_impact['First Category'] == 'Technology', 'Avg Lifetime Profit'].values[0]
difference = ((furniture_customers - tech_customers) / tech_customers * 100)
print(f"\n🚨 STRATEGIC FINDING:")
print(f"   Customer who first buy Furniture:")
print(f"   - Have {difference:.1f}% {'LOWER' if difference < 0 else 'HIGHER'} lifetime profit")
print(f"   - Average lifetime profit: ${furniture_customers:.2f}")
print(f"\n   Customer who first buy Technology:")
print(f"   - Average liftime profit: ${tech_customers:.2f}")
print(f"\n💡 STRATEGY: Offer Technology products as first-purchase incentives")

# Visualization
fig = px.bar(first_category_impact,
             x = 'First Category',y = 'Avg Lifetime Profit',
             color = 'Avg Lifetime Profit', color_continuous_scale = 'RdYlGn',
             title = '🎯 Customer Lifetime Value by First Purchase Category')
fig.update_layout(height = 500, template = 'plotly_white')
fig.show()


🎯 FIRST PURCHASE CATEGORY IMPACT ON CUSTOMER LIFETIME VALUE
    First Category  Avg Lifetime Sales  Avg Lifetime Profit  Avg Orders  \
0        Furniture         3041.963313           357.785161    6.305882   
1  Office Supplies         2700.928522           333.892422    6.207171   
2       Technology         3505.793214           479.004534    6.785124   

   Avg Discount  Customer Count  
0      0.163838             170  
1      0.157655             502  
2      0.147834             121  

🚨 STRATEGIC FINDING:
   Customer who first buy Furniture:
   - Have -25.3% LOWER lifetime profit
   - Average lifetime profit: $357.79

   Customer who first buy Technology:
   - Average liftime profit: $479.00

💡 STRATEGY: Offer Technology products as first-purchase incentives


### Finding #4. Loss Leaders That Actually Work
### Some unprofitable products drive future profitable purchases

In [35]:
# For customer who bought unprofitable products
unprofitable_products = df[df['Profit'] < 0].groupby('Product Name').agg({
    'Profit': 'sum',
    'Order ID': 'count',
    'Customer ID': 'nunique'
}).reset_index()
unprofitable_products.columns = ['Product Name', 'Total Loss', 'Times Sold', 'Unique Customers']
unprofitable_products = unprofitable_products.sort_values('Total Loss').head(10)

# For each unprofitable product, find if customers came back
loss_leader_analysis = []
for product in unprofitable_products['Product Name'].head(5):
    customers_bought = df[df["Product Name"] == product]['Customer ID'].unique()

    # Their subsequent orders (after buying this product)
    subsequent_orders = df[
        (df['Customer ID'].isin(customers_bought)) &
        (df['Product Name'] != product)
    ]
    if len(subsequent_orders) > 0:
        subsequent_profit = subsequent_orders.groupby('Customer ID')['Profit'].sum().mean()
        subsequent_orders_count = subsequent_orders.groupby('Customer ID').size().mean()
        loss_leader_analysis.append({
            'Product': product[:30] + '...' if len(product) > 30 else product,
            'Initial Loss': df[df['Product Name'] == product]['Profit'].sum(),
            'Customers': len(customers_bought),
            'Avg Future Orders': subsequent_orders_count,
            'Avg Future Profit': subsequent_profit,
            'Net Impact': (subsequent_profit * len(customers_bought)) + df[df['Product Name'] == product]['Profit'].sum()
        })
loss_leader_df = pd.DataFrame(loss_leader_analysis)

print("\n" + "=" * 80)
print("💰 LOSS LEADER ANALYSIS: Products That Lose Money But Build Relationships")
print("=" * 80)
print(loss_leader_df.round(2))

if len(loss_leader_df[loss_leader_df['Net Impact'] > 0]) > 0:
    print(f"\n🎯 GAME-CHANGING INSIGHT:")
    print(f"   {len(loss_leader_df[loss_leader_df['Net Impact'] > 0])} 'unprofitable' products actually:")
    print(f"   - Attract customers who make {loss_leader_df['Avg Future Orders'].mean():.1f} more purchases")
    print(f"   - Generate ${loss_leader_df['Avg Future Profit'].mean():.2f} in future profit per customer")
    print(f"   - Net positive impact: ${loss_leader_df[loss_leader_df['Net Impact'] > 0]['Net Impact'].sum():.2f}")
    print(f"\n💡 RECOMMENDATION: KEEP these products as strategic loss leaders")

#Visualize
fig = px.scatter(loss_leader_df,
                 x='Initial Loss',
                 y='Avg Future Profit',
                 size='Customers',
                 hover_data=['Product'],
                 color='Net Impact',
                 color_continuous_scale='RdYlGn',
                 title='💰 Loss Leader Effectiveness: Initial Loss vs Future Gain')
fig.add_hline(y=0, line_dash="dash", line_color="gray")
fig.add_vline(x=0, line_dash="dash", line_color="gray")
fig.update_layout(height=500, template='plotly_white')
fig.show()


💰 LOSS LEADER ANALYSIS: Products That Lose Money But Build Relationships
                             Product  Initial Loss  Customers  \
0  Cubify CubeX 3D Printer Double...      -8879.97          3   
1  GBC DocuBind P400 Electric Bin...      -1878.17          6   
2  Lexmark MX611dhe Monochrome La...      -4589.97          4   
3  GBC Ibimaster 500 Manual ProCl...        760.98          9   
4  GBC DocuBind TL300 Electric Bi...       2233.51         11   

   Avg Future Orders  Avg Future Profit  Net Impact  
0              11.67             624.12    -7007.60  
1              15.17             467.62      927.53  
2              13.75            -391.35    -6155.36  
3              11.22            -562.18    -4298.63  
4              17.27            1004.93    13287.78  

🎯 GAME-CHANGING INSIGHT:
   2 'unprofitable' products actually:
   - Attract customers who make 13.8 more purchases
   - Generate $228.63 in future profit per customer
   - Net positive impact: $14215.32

💡 REC

### Insights#5: The Holdiay Sales Profitability Trap

In [37]:
# Add markdown:
"""
### Finding #5: The Holiday Sales Profitability Trap
"""

# Quarterly analysis with deeper metrics
quarterly_analysis = df.groupby(['Year', 'Quarter']).agg({
    'Sales': 'sum',
    'Profit': 'sum',
    'Discount': 'mean',
    'Order ID': 'count',
    'Is Profitable': 'mean'
}).reset_index()

quarterly_analysis['Profit Margin %'] = (quarterly_analysis['Profit'] / quarterly_analysis['Sales'] * 100).round(2)
quarterly_analysis['Avg Order Value'] = (quarterly_analysis['Sales'] / quarterly_analysis['Order ID']).round(2)
quarterly_analysis['Profit Rate %'] = (quarterly_analysis['Is Profitable'] * 100).round(2)

print("\n" + "=" * 100)
print("📅 QUARTERLY PERFORMANCE: The Revenue vs Profit Paradox")
print("=" * 100)
print(quarterly_analysis)

# Find Q4 pattern
q4_data = quarterly_analysis[quarterly_analysis['Quarter'] == 4]
other_q = quarterly_analysis[quarterly_analysis['Quarter'] != 4]

q4_margin = q4_data['Profit Margin %'].mean()
other_margin = other_q['Profit Margin %'].mean()

print(f"\n🎄 Q4 (Holiday Season) Analysis:")
print(f"   - Average Profit Margin: {q4_margin:.2f}%")
print(f"   - Average Discount: {q4_data['Discount'].mean():.2%}")
print(f"\n📊 Other Quarters:")
print(f"   - Average Profit Margin: {other_margin:.2f}%")
print(f"   - Average Discount: {other_q['Discount'].mean():.2%}")

margin_diff = ((q4_margin - other_margin) / other_margin * 100)
print(f"\n🚨 CRITICAL INSIGHT:")
print(f"   Q4 profit margins are {abs(margin_diff):.1f}% {'LOWER' if margin_diff < 0 else 'HIGHER'} than other quarters")
print(f"   despite {((q4_data['Sales'].sum() / other_q['Sales'].mean() - 1) * 100):.1f}% higher sales volume")
print(f"\n💡 STRATEGY: Reduce Q4 discounts from {q4_data['Discount'].mean():.1%} to {other_q['Discount'].mean():.1%}")
print(f"   Projected profit increase: ${(q4_data['Sales'].sum() * (other_margin - q4_margin) / 100):.2f}")

# Visualize
fig = go.Figure()
fig.add_trace(go.Bar(name='Sales', x=quarterly_analysis['Quarter'], y=quarterly_analysis['Sales'], yaxis='y', offsetgroup=1))
fig.add_trace(go.Scatter(name='Profit Margin %', x=quarterly_analysis['Quarter'], y=quarterly_analysis['Profit Margin %'],
                         yaxis='y2', mode='lines+markers', line=dict(color='red', width=3), marker=dict(size=10)))

fig.update_layout(
    title='📅 The Q4 Paradox: High Sales, Low Margins',
    xaxis=dict(title='Quarter'),
    yaxis=dict(title='Total Sales ($)', side='left'),
    yaxis2=dict(title='Profit Margin %', overlaying='y', side='right'),
    template='plotly_white',
    height=500,
    hovermode='x unified'
)
fig.show()


📅 QUARTERLY PERFORMANCE: The Revenue vs Profit Paradox
    Year  Quarter        Sales      Profit  Discount  Order ID  Is Profitable  \
0   2011        1   74447.7960   3811.2290  0.157447       282       0.794326   
1   2011        2   86538.7596  11204.0692  0.145459       392       0.831633   
2   2011        3  143633.2123  12804.7218  0.155089       564       0.820922   
3   2011        4  179627.7302  21723.9541  0.167642       755       0.776159   
4   2012        1   68851.7386   9264.9416  0.146000       260       0.807692   
5   2012        2   89124.1870  12190.9224  0.171622       444       0.806306   
6   2012        3  130259.5752  16853.6194  0.152061       592       0.793919   
7   2012        4  182297.0082  23309.1203  0.152494       806       0.816377   
8   2013        1   92596.4190  11446.3415  0.151265       332       0.783133   
9   2013        2  135370.1130  16084.9055  0.161785       594       0.803030   
10  2013        3  144614.4282  16153.5046  0.159460 